### ISSUE TITLE
Standalone generated quantities following optimization or VI

### ISSUE THREAD

**Issue by WardBrian** [stan-dev/cmdstan#1099]
#### Summary:
Currently, the standalone generate_quantities function only accepts a `.csv` file from sampling. It would be nice to present a consistent interface which allowed the use of any Stan CSV file. The docs seem to imply this would be the intended behavior, since they never specify that the fitted_params argument must be a HMC run.

#### Description:

Right now, one can have a reasonable workflow using two models, one which fits params, and another which predicts for new data. However, this workflow does not translate when swapping the algorithm between sampling, optimization, and VI. 

For example, the following two models:

```stan
data {
  int<lower=0> N;
  vector[N] x;
  vector[N] y;
}
parameters {
  real alpha;
  real beta;
  real<lower=0> sigma;
}
model {
  y ~ normal(alpha + beta * x, sigma);
}
```

```stan
data {
  int<lower=0> N;
  vector[N] x_new;
}
parameters {
  real alpha;
  real beta;
  real<lower=0> sigma;
}
generated quantities {
  array[N] real y_new = normal_rng(alpha + beta * x_new, sigma);
}
```

If you are targetting this kind of two-step workflow, using separate models and standalone GQ seems like the most natural fit, however, this currently only works if you run sampling as the `fit` step of the fit/predict paradigm. 

The alternatives that come to mind are either using tricks like zero-size objects to do this all in one model while still keeping them as separate steps, or doing the predictive sampling downstream in Python/R.

#### Current Output:

Using an output.csv from optimization with the `generate_quantities` function produces an unhelpful error:
```
Mismatch between model and fitted_parameters csv file "output.csv"
```

#### Expected Output:
I'd expect the generated quantities to be computed based on the parameter estimates from optimization or VI.

#### Additional Information:
Provide any additional information here.

#### Current Version:
v2.29.2

**Comment by WardBrian** [comment]
I should note that standalone generated quantities does work fine when the number of samples is `1`, so I see no technical reason why the result from optimization couldn't be used in the same manner

**Comment by betanalpha** [comment]
> I should note that standalone generated quantities does work fine when the number of samples is 1, so I see no technical reason why the result from optimization couldn't be used in the same manner.

The nomenclature that has become conventional obscures things a bit, but mathematically “standalone generated quantities” should be applicable to any method that returns some collection of points from the model configuration space, which currently includes all of the current methods.

Just so that we’re all on the same page: if Theta is the model configuration space defined in the `parameters` block and Y is the observational space defined in the `data and `transformed data` blocks then the `generated quantities` block just defines a function Theta x Y -> Z, where Z is some useful auxiliary space.  Each method then implicitly applies this function to certain points (theta, y) \in Theta x Y; for example  by default the MCMC implementations apply the function to Markov chain states generated after the initial warmup period and the optimization implementation applies the function to the final optimum.  One complication is that these output points have never really been explicitly defined, but only implicitly defined by the method implementations.

To generalize standalone generated quantities to any method I think it would help to introduce a more general vocabulary for “points returned by a given method that are then used for probabilistic computation”; for example “active states” or something.  Then standalone generated quantities could be generally defined as defining a function as above and then evaluated it on the active states saved in the output of a previous fit., and the implementation would be unambiguous once each method explicitly defined its output.

The latter step would require some choices — for example if optimization is run in verbose mode so that the full optimization path is saved should the generated quantities function be applied to all of those states, or just the final state? — but it would compartmentalize those degrees of freedom from the standalone generated quantities design and implementation.

**Comment by WardBrian** [comment]
> should be applicable to any method that returns some collection of points from the model configuration space, which currently includes all of the current methods.

Exactly. Getting the terminology for this just right might be tricky, but it think it would be worthwhile to do so

A good point on optimization - my first instinct is that this should be from the final iteration only in that case. That shouldn’t be a problem in terms of implementation or documentation however

**Comment by betanalpha** [comment]
> Exactly. Getting the terminology for this just right might be tricky, but it think it would be worthwhile to do so
> 
> A good point on optimization - my first instinct is that this should be from the final iteration only in that case. That shouldn’t be a problem in terms of implementation or documentation however
> 
Agreed — most of the confusion is due to implicit definitions and not any technical hurdles.

The CmdStan output was designed to be relatively agnostic — header metadata, intermediate comments to accommodate updated information such as the results of adaptation, footer metadata, and then state information in between.  The state information could be a single state or an ensemble of states, with the ensemble having multiple possible interpretations (correlated states in a Markov chain for HMC, uncorrelated states sampled from the variational solution in ADVI, optimization path in LBFGS).  Because of this design any implementation of standalone generated quantities can just discard the header, footer, and intermediate metadata and then apply the function defined in the new Stan program to each of the states or some subset of the states.  Uniformly applying the function to all of the states would be easiest to implement, but method-specific choices based on the header metadata are also possible.

**Comment by WardBrian** [comment]
I just checked, and it looks like if you save the warmup draws in HMC, the standalone generated quantities will also run with those warmup draws. At the python level in cmdstanpy we wrap this in such a way where warmup is excluded unless you request it.

I think it's fair enough to say we want standalone GQ to work generically over all our outputs and does something for each and every row found in them, rather than needing to figure out a subset based on the original run's metadata.
